### PR TITLE
Bug 1213355 - GaiaHeader.go_back shows ElementNotVisibleException

### PR DESCRIPTION
**Pull request by nojunpark**

### CODE REVIEW

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=6aed75f235e76d33a075087ed2fdcfd4a93d5eef

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=3d620a2f70709fa2417ee1e982c35d73786df6be

File: tests/python/gaia-ui-tests/gaiatest/form_controls/form_control.py [hunk 1]
```diff
@@ -8,5 +8,8 @@
 
 class Widget(PageRegion):
     def __init__(self, marionette, locator):
-        element = Wait(marionette).until(expected.element_present(*locator))
+        if type(locator) is tuple:
```
Comment (JohanLorenzo) [inline tests/python/gaia-ui-tests/gaiatest/form_controls/form_control.py]: I agree, that would be the way we do in JavaScript. But in python, we can avoid to force a parameter in lieu of another:
```python
def __init__(self, marionette, locator=None, element=None):
   if locator is None and element is None:
        raise AttributeError('Either the locator or the element should be provided')
   if element is None:
        element =  Wait(marionette).until(expected.element_present(*locator))
   PageRegion.__init__(self, marionette, element)
```

File: tests/python/gaia-ui-tests/gaiatest/apps/settings/app.py [hunk 2]
```diff
@@ -339,7 +340,7 @@ def _wait_for_toggle_ready(self, by, locator):
         Wait(self.marionette).until(expected.element_enabled(checkbox))
 
     def return_to_prev_menu(self, parent_view, exit_view):
-        GaiaHeader(self.marionette, self._header_locator).go_back()
+        GaiaHeader(self.marionette, exit_view.find_element(*self._header_locator)).go_back()
```
Comment (JohanLorenzo) [inline tests/python/gaia-ui-tests/gaiatest/apps/settings/app.py]: What is the difference between the old and the new line? If I understand the `GaiaHeader` constructor correctly, we're passing the element instead of the locator, right?

File: tests/python/gaia-ui-tests/gaiatest/apps/settings/regions/homescreen_settings.py [hunk 3]
```diff
@@ -16,8 +16,9 @@ class HomescreenSettings(Base):
 
     _wallpaper_preview_locator = (By.CSS_SELECTOR, '.wallpaper-preview')
     _wallpaper_pick_locator = (By.CLASS_NAME, 'wallpaper-button')
-    _pick_cancel_button_locator = (By.CSS_SELECTOR,
-                                   '#action-menu-list > button[data-l10n-id="cancel"]')
+    _pick_cancel_button_locator = (By.CSS_SELECTOR, '#action-menu-list > '
```
Comment (JohanLorenzo) [inline tests/python/gaia-ui-tests/gaiatest/apps/settings/regions/homescreen_settings.py]: This selector can be reduced to `'#action-menu-list button[data-l10n-id="cancel"]` if we get rid of the child operators. The dom below `#action-menu-list` is small enough so the collision between 2 potentials `button[data-l10n-id="cancel"]` is low.

Comment (npark-mozilla) [inline tests/python/gaia-ui-tests/gaiatest/apps/settings/regions/homescreen_settings.py]: Strange, I changed from '#action-menu-list > button[data-l10n-id="cancel"]' because that one wasn't working.  I thought it is almost identical to '#action-menu-list button[data-l10n-id="cancel"]', I guess it only searches for the direct child, not its descendants?

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=2cad7d94b60057e6eb285a6fd89441e4cf819e0a

File: tests/python/gaia-ui-tests/gaiatest/apps/settings/app.py [hunk 4]
```diff
@@ -12,6 +12,7 @@ class Settings(Base):
     name = 'Settings'
 
     _header_locator = (By.CSS_SELECTOR, '.current gaia-header')
+    _html_header_locator = (By.CSS_SELECTOR, '.current header')
```
Comment (mwargers) [inline tests/python/gaia-ui-tests/gaiatest/apps/settings/app.py]: Is this still needed? It doesn't seem used anymore in your pull request.

File: tests/python/gaia-ui-tests/gaiatest/apps/settings/app.py (hunk 2 above)
Comment (mwargers) [inline tests/python/gaia-ui-tests/gaiatest/apps/settings/app.py]: I believe this is to make sure that the right element is used (child of exit_view).

File: tests/python/gaia-ui-tests/gaiatest/apps/system/regions/fxaccounts.py [hunk 5]
```diff
@@ -68,6 +70,7 @@ def wait_for_successful_login(self):
         return done
 
     def tap_done(self):
+        self._switch_to_fxa_iframe()
         done = self.wait_for_successful_login()
         done.tap()
         #Switch back to the settings app
```
Comment (mwargers) [inline tests/python/gaia-ui-tests/gaiatest/apps/system/regions/fxaccounts.py]: The changes in this file seem totally unrelated to the bug, no?

File: tests/python/gaia-ui-tests/gaiatest/form_controls/form_control.py (hunk 1 above)
Comment (mwargers) [inline tests/python/gaia-ui-tests/gaiatest/form_controls/form_control.py]: You need to do this, still.

File: tests/python/gaia-ui-tests/gaiatest/apps/settings/app.py (hunk 4 above)
Comment (npark-mozilla) [inline tests/python/gaia-ui-tests/gaiatest/apps/settings/app.py]: removed

File: tests/python/gaia-ui-tests/gaiatest/apps/system/regions/fxaccounts.py (hunk 5 above)
Comment (npark-mozilla) [inline tests/python/gaia-ui-tests/gaiatest/apps/system/regions/fxaccounts.py]: Correct, I was thinking about creating a separate bug for fxa page related changes, but I thought changes were small enough to be added to this PR.  (And without this changes, I wouldn't know whether the return_to_prev_menu() method is working properly in fxa page, since it works within its own iframe)  Will try to separate the works in the future.

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=eb39e2bd03d3a2a4aef90cc0ac8373023078af45

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=7208e40945f4b7f4ae73d3295671f67541c80171

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=00e00b9ac347f642c98db18856ad3f4f76143f5f

File: tests/python/gaia-ui-tests/gaiatest/apps/system/regions/fxaccounts.py (hunk 5 above)
Comment (mwargers) [inline tests/python/gaia-ui-tests/gaiatest/apps/system/regions/fxaccounts.py]: It's fine, but can you mention it when you ask review. I was confused by this change.

File: tests/python/gaia-ui-tests/gaiatest/apps/settings/regions/homescreen_settings.py (hunk 3 above)
Comment (JohanLorenzo) [inline tests/python/gaia-ui-tests/gaiatest/apps/settings/regions/homescreen_settings.py]: Yup, `>` is the direct child operator
